File: merge_sort.py
'''
A standard divide and conquer algorithm for mergesort

input format
============
12 11 13 5 6 7
==============
'''

arr = list(map(int, input().split(" ")))

def merge(a,b):
	l = []
	while a and b:
		if a[0] < b[0]:
			l.append(a.pop(0))
		else:
			l.append(b.pop(0))
	return l + a + b


def merge_sort(l, h):
	mid = int((h+l)/2)
	if l < h:
		# print(arr[l: h])
		low = merge_sort(l, mid)
		high = merge_sort(mid+1, h)
		return merge(low, high)
	# print(arr, l)
	return [arr[l]]
print(merge_sort(0, len(arr)-1))

File: merge_sort.cpp
/*
A standard divide and conquer algorithm for mergesort

input format
============
12 11 13 5 6 7
==============
*/
#include <bits/stdc++.h>
std::vector<int> v, answer;

std::vector<int> merge(std::vector<int> low, std::vector<int> high){
	std::vector<int> l, ret;
	while(!low.empty() && !high.empty()){
		if(low[0] < high[0]){
			l.push_back(low[0]);
			low.erase(low.begin());
		}
		else{
			l.push_back(high[0]);
			high.erase(high.begin());
		}
	}
	ret.reserve(l.size() + low.size() + high.size());
	ret.insert(ret.end(), l.begin(), l.end());
	ret.insert(ret.end(), low.begin(), low.end());
	ret.insert(ret.end(), high.begin(), high.end());
	return ret;
}

std::vector<int> mergesort(int l, int h){
	std::vector<int> low;
	std::vector<int> high;
	int mid = (l+h)/2;
	if (l < h){
		low = mergesort(l, mid);
		high = mergesort(mid+1, h);
		return merge(low, high);
	}
	return {v[l]};
}

int main(){
	int temp;
	for(int i = 0; i<6; i++){
		std::cin >> temp;
		v.push_back(temp);
	}

	answer = mergesort(0, 5);
	for(int i = 0; i < 6; i++){
		std::cout << answer[i] << " "; 
	}
	return 0;
}